src/lib/core/decimal.h

```diff
@@ -37,52 +37,151 @@
 #include <decNumber/decNumber.h>
 #include <stdint.h>
 #include <stdbool.h>
+#include "trivia/util.h"
 
 #if defined(__cplusplus)
 extern "C" {
 #endif /* defined(__cplusplus) */
 
-enum {
-	/*
-	 * The documentation in the decnumber sources says the string needs to
-	 * be >= digit count + 14.
+/*
+ * We should alias <decNumber> to <decimal_t> within tarantool
+ * sources.
+ *
+ * The type is defined in decNumber.h using an unnamed structure:
+ *
+ * typedef struct {
+ *         <...>
+ * } decNumber;
+ */
+#define STRUCT_DECIMAL decNumber
+
+/** \cond public */
+
+#define DECIMAL_MAX_STR_LEN 52
+
+/**
+ * The structure behind <decimal_t> is transparent for a user
+ * and named <struct decimal> by default.
+ *
+ * Define STRUCT_DECIMAL macro before inclusion of the module.h
+ * header to alias <decimal_t> to another structure. It should
+ * have the same data layout as one below: exactly same size (36)
+ * and alignment (4).
+ */
+#ifndef STRUCT_DECIMAL
+
+/**
+ * Decimal number.
+ *
+ * The data layout is the same as in the decNumber library.
+ */
+struct decimal {
```

Comment:
I don't like exposing the decimal number layout to the user. If we switch to a different decimal library or even just reconfigure libdecnumber, as @sergepetrenko pointed out, we'll break the ABI. We can avoid potential ABI breakages by reserving a substantial amount of bytes for the decimal object:

```C
struct decimal {
    char data[64];
};
```

Reply:
This way the following code (which uses an API from PR #7429) will write behind the buffer:

```c
int
luaT_push_twice(struct lua_State *L, const decimal_t *dec)
{
        decimal_t *res_1 = luaT_newdecimal(L);
        decimal_t *res_2 = luaT_newdecimal(L);
        *res_1 = *dec;
        *res_2 = *dec;
        return 2;
}
```

*(I would discuss both C and Lua/C API verbally, that's tricky.)*

Reply:
I guess we could add `decimal_copy` to avoid that.